File: data-structures/javascript/hash-map.js
const assert = require('assert')

class LinkedList {
  constructor(data) {
    this.data = data;
    this.next = null;
  }
}

function adler32(text) {
  let s1 = 1;
  let s2 = 0;

  for (let i = 0; i < text.length; i++) {
    let n = text.charCodeAt(i);
    s1 = (s1 + n) % 65521;
    s2 = (s2 + s1) % 65521;
  }

  return (s2 << 16) | s1;
}

const dbg = (label, data) => {
  if (data == null) {
    data = label;
    label = null;
  }

  if (label == null) {
    console.log(data);
  } else {
    console.log(label, data);
  }

  return data;
}

const BASE_CAP = 500;

class HashMap {
  constructor() {
    this.capacity = BASE_CAP;
    this.buckets = new Array(this.capacity).fill(null);
  }

  get(key) {
    let bucket = this.hash(key) % this.buckets.length;
    let list = this.buckets[bucket];
    if (list) {
      let node = list;
      while (node) {
        if (String(node.data.key) === String(key)) {
          return node.data.value;
        }
        node = node.next;
      }
    }
    return null;
  }

  set(key, value) {
    let bucket = this.hash(key) % this.buckets.length;
    if (this.buckets[bucket]) {
      let node = this.buckets[bucket];
      while (node.next) {
        node = node.next;
      }
      node.next = new LinkedList({ key, value });
    } else {
      this.buckets[bucket] = new LinkedList({ key, value });
    }
  }

  hash(key) {
    return adler32(String(key))
  }
}

let map = new HashMap()

assert.deepEqual(map.get(2), null)

map.set(2, 50)

assert.deepEqual(map.get(2), 50)

map.set(2, 60)

assert.deepEqual(map.get(2), 60)

File: data-structures/rust/src/hash_map.rs
fn adler32(text: &str) -> u32 {
  let mut s1 = 1;
  let mut s2 = 0;

  for c in text.chars() {
    let n = u32::from(c);
    s1 = (s1 + n) % 65521;
    s2 = (s2 + s1) % 65521;
  }

  return (s2 << 16) | s1;
}

#[derive(Debug)]
struct LinkedList<T> {
    data: T,
    next: Option<Box<LinkedList<T>>>,
}

impl<T> LinkedList<T> {
    fn new(data: T) -> Self {
        Self {
            data,
            next: None,
        }
    }
}

const BASE_CAP: usize = 500;

#[derive(Debug)]
pub struct HashMap<'a, V> {
    entries: [Option<Box<LinkedList<(&'a str, V)>>>; BASE_CAP],
}

impl<'a, V> HashMap<'a, V> {
    pub fn new() -> Self {
        Self { entries: std::array::from_fn(|_| None) }
    }

    pub fn get(&self, key: &str) -> Option<&V> {
        let bucket = adler32(key) as usize % BASE_CAP;
        let mut list = self.entries[bucket].as_ref();
        while let Some(node) = list {
            if node.data.0 == key {
                return Some(&node.data.1);
            } else {
                list = node.next.as_ref();
            }
        }
        None
    }

    pub fn set(&mut self, key: &'a str, value: V) {
        let bucket = adler32(key) as usize % BASE_CAP;
        let mut list = self.entries[bucket].as_mut();
        if list.is_some() {
            while let Some(node) = list {
                list = node.next.as_mut();
            }
            list.replace(&mut Box::new(LinkedList::new((key, value))));
        } else {
            self.entries[bucket] = Some(Box::new(LinkedList::new((key, value))));
        }
    }
}

#[test]
fn test_map() {
    let mut map = HashMap::new();
    assert_eq!(map.get("2"), None);

    map.set("2", 50);
    assert_eq!(map.get("2"), Some(&50));

    map.set("2", 60);
    assert_eq!(map.get("2"), Some(&60));

    dbg!(map);
}
